### PR TITLE
Keras Resnet Model for the model garden

### PR DESCRIPTION
**Pull request by anj-s**
Added a Keras Resnet model that works with tf.Dataset

### CODE REVIEW

File: official/keras/cifar10_dataset.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import keras
```
Comment (fchollet) [inline official/keras/cifar10_dataset.py]: import `from tensorflow import keras` instead

File: official/keras/cifar10_dataset.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+from keras.datasets import cifar10
+import numpy as np
+import tensorflow as tf
+
+def get_train_eval_dataset():
```
Comment (fchollet) [inline official/keras/cifar10_dataset.py]: It would be preferable to use something from core TF for CIFAR10. Do we have it in TF?

File: official/keras/resnet50_cifar10_model.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
+from __future__ import print_function
+
+import keras
+from keras import backend as K
+from keras.engine.topology import get_source_inputs
+from keras.layers import AveragePooling2D, Input, Flatten
+from keras.layers import Dense, Conv2D, BatchNormalization, Activation
+from keras.models import Model
```
Comment (fchollet) [inline official/keras/resnet50_cifar10_model.py]: Prefer importing modules for user-facing code:

```python
import keras
from keras import layers
```

These two are basically all you need in this file.

Also: one import per line

File: official/keras/resnet50_cifar10_model.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+# Model version
+# Orig paper: version = 1 (ResNet v1), Improved ResNet: version = 2 (ResNet v2)
+version = 1
+
+# Computed depth from supplied model parameter n
+depth = n * 6 + 2
+
+# Model name, depth and version
+model_type = 'ResNet%dv%d' % (depth, version)
```
Comment (fchollet) [inline official/keras/resnet50_cifar10_model.py]: Global constant can be in the top-level namespace, but should be ALL_CAPS.

File: official/keras/resnet50_cifar10_model.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+def resnet_block(inputs,
+    num_filters=16,
+    kernel_size=3,
+    strides=1,
+    activation='relu',
+    conv_first=True):
```
Comment (fchollet) [inline official/keras/resnet50_cifar10_model.py]: Fix indent

File: official/keras/resnet50_cifar10_model.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  """2D Convolution-Batch Normalization-Activation stack builder
+  # Arguments
```
Comment (fchollet) [inline official/keras/resnet50_cifar10_model.py]: Introduce blank line above

File: official/keras/resnet50_cifar10_model.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+      inputs (tensor): input tensor from input image or previous layer
+      num_filters (int): Conv2D number of filters
+      kernel_size (int): Conv2D square kernel dimensions
+      strides (int): Conv2D square stride dimensions
+      activation (string): activation name
+      conv_first (bool): conv-bn-activation (True) or
+          activation-bn-conv (False)
+  # Returns
```
Comment (fchollet) [inline official/keras/resnet50_cifar10_model.py]: Introduce blank line above

File: official/keras/resnet50_cifar10_model.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+             strides=strides,
+             padding='same',
+             kernel_initializer='he_normal',
+             kernel_regularizer=l2(1e-4))(x)
+  return x
+
+
+def resnet_v1(input_shape, depth, num_classes=10,input_tensor=None):
+  """ResNet Version 1 Model builder [a]
+  Stacks of 2 x (3 x 3) Conv2D-BN-ReLU
```
Comment (fchollet) [inline official/keras/resnet50_cifar10_model.py]: Introduce blank line above

File: official/keras/resnet50_cifar10_model.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+  Last ReLU is after the shortcut connection.
+  The number of filters doubles when the feature maps size
+  is halved.
+  The Number of parameters is approx the same as Table 6 of [a]:
+  ResNet20 0.27M
+  ResNet32 0.46M
+  ResNet44 0.66M
+  ResNet56 0.85M
+  ResNet110 1.7M
+  # Arguments
```
Comment (fchollet) [inline official/keras/resnet50_cifar10_model.py]: Introduce blank line above

File: official/keras/resnet50_cifar10_model.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+      input_shape (tensor): shape of input image tensor
+      depth (int): number of core convolutional layers
+      num_classes (int): number of classes (CIFAR10 has 10)
+  # Returns
```
Comment (fchollet) [inline official/keras/resnet50_cifar10_model.py]: Introduce blank line above

File: official/keras/resnet50_cifar10_model.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+  x = AveragePooling2D(pool_size=8)(x)
+  y = Flatten()(x)
+  outputs = Dense(num_classes,
+                  activation='softmax',
+                  kernel_initializer='he_normal')(y)
+
+  # Ensure that the model takes into account
+  # any potential predecessors of `input_tensor`.
+  if input_tensor is not None:
+    inputs = get_source_inputs(input_tensor)
```
Comment (fchollet) [inline official/keras/resnet50_cifar10_model.py]: You can remove this block, `img_input` will always work in this context (also `get_source_inputs` is not part of the public API).

File: official/keras/train.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+  checkpoint = ModelCheckpoint(filepath=filepath,
+                               monitor='val_acc',
+                               verbose=1,
+                               save_best_only=True)
+  callbacks = [checkpoint]
+  model.fit(steps_per_epoch=1, epochs=FLAGS.epochs, verbose=1,
+            callbacks=callbacks)
+  model.save_weights('saved_wt.h5')
+
+  # Evaluate model
```
Comment (fchollet) [inline official/keras/train.py]: Please put evaluation in a separate function

File: official/keras/train.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+  test_dataset = test_dataset.batch(FLAGS.batch_size)
+
+  (train_images, train_labels) = train_dataset.make_one_shot_iterator().get_next()
+  train_images = tf.cast(train_images, tf.float32)
+  train_labels = tf.cast(train_labels, tf.float32)
+  (test_images, test_labels) = test_dataset.make_one_shot_iterator().get_next()
+  test_images = tf.cast(test_images, tf.float32)
+  test_labels = tf.cast(test_labels, tf.float32)
+
+  model = resnet50_model(train_images)
```
Comment (fchollet) [inline official/keras/train.py]: Before defining the model, set the learning phase to `True` (and maybe set `image_data_format` to `channels_first`, or make it configurable)

File: official/keras/train.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+  model.compile(loss=loss_fn,
+                 optimizer=keras.optimizers.TFOptimizer(optimizer),
```
Comment (fchollet) [inline official/keras/train.py]: Fix indent

Also: you can just pass `optimizer= optimizer`

File: official/keras/resnet50_cifar10_model.py [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+               kernel_size=kernel_size,
+               strides=strides,
+               padding='same',
+               kernel_initializer='he_normal',
+               kernel_regularizer=l2(1e-4))(inputs)
+    x = BatchNormalization()(x)
+    if activation:
+      x = Activation(activation)(x)
+    return x
+  x = BatchNormalization()(inputs)
```
Comment (fchollet) [inline official/keras/resnet50_cifar10_model.py]: Introduce blank line above

File: official/keras/resnet50_cifar10_model.py [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+      x = keras.layers.add([x, y])
+      x = Activation('relu')(x)
+    num_filters = 2 * num_filters
+
+  # Add classifier on top.
+  # v1 does not use BN after last shortcut connection-ReLU
+  x = AveragePooling2D(pool_size=8)(x)
+  y = Flatten()(x)
+  outputs = Dense(num_classes,
+                  activation='softmax',
```
Comment (fchollet) [inline official/keras/resnet50_cifar10_model.py]: We should consider not including the softmax in the training model (and use `softmax_crossentropy` as loss), for performance reasons 

Depends if performance is critical or not

File: official/keras/train.py [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+  # Print eval results
+  print('Test loss:', loss)
+  print('Test accuracy:', acc)
+
+
+def resnet50_model(train_images_tensors):
+  depth = 50
+  input_shape = (32, 32, 3)
+  input_vals = keras.layers.Input(tensor=train_images_tensors)
+  model = resnet50_cifar10_model.resnet_v1(input_shape=input_shape, depth=depth,
```
Comment (fchollet) [inline official/keras/train.py]: One keyword argument per line

File: official/keras/train.py [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+                                           input_tensor=input_vals)
+  return model
+
+def main(unparser):
+  input_shape, train_dataset, test_dataset = cifar10_dataset.get_train_eval_dataset()
+
+  train(resnet50_cifar10_model, train_dataset, test_dataset,
+        tf.train.AdamOptimizer(learning_rate=0.001), loss_fn='categorical_crossentropy')
```
Comment (fchollet) [inline official/keras/train.py]: One keyword argument per line

File: official/keras/cifar10_dataset.py (hunk 2 above)
Comment (anj-s) [reply official/keras/cifar10_dataset.py]: Yes, I agree. This was one of the things I was planning on refactoring. I found an example in the model repo that converts the cifar10 dataset to TFRecordDataset. We probably need to write a new one.

File: official/keras/resnet50_cifar10_model.py (hunk 16 above)
Comment (anj-s) [reply official/keras/resnet50_cifar10_model.py]: Performance is critical since we will be benchmarking this models. However I think it will not be this particular model. I made the change anyway and will keep this in mind going forward.
